### ISSUE TITLE
Allow to create sublibraries named "opam"

### ISSUE THREAD

**Issue by kit-ty-kate** [ocaml/dune#4068]
## Expected Behavior

Be able to have a sublibrary named `opam` (e.g. `<pkg>.opam`)

## Actual Behavior & Reproduction

```
$ cat dune
(library
 (name test)
 (public_name test.opam))
$ touch test.ml test.opam
$ dune build
Error: symlink: _build/install/default/lib/test/opam/test.ml: Not a directory
Error: symlink: _build/install/default/lib/test/opam/test.cmi: Not a
directory
Error: symlink: _build/install/default/lib/test/opam/test.cmt: Not a
directory
Error: symlink: _build/install/default/lib/test/opam/test.cma: Not a
directory
Error: symlink: _build/install/default/lib/test/opam/test.cmx: Not a
directory
Error: symlink: _build/install/default/lib/test/opam/test.a: Not a directory
Error: symlink: _build/install/default/lib/test/opam/test.cmxa: Not a
directory
Error: symlink: _build/install/default/lib/test/opam/test.cmxs: Not a
directory
```

## Specifications

- Version of `dune` (output of `dune --version`): 2.7.1

**Comment by josch** [comment]
FWIW this issue is not just hypothetical. We are currently trying to use dune for [dose3](https://opam.ocaml.org/packages/dose3/dose3.5.0.1/) and since dose3 also supports opam, it has a submodule named like that. Switching the dose3 build system to dune is currently blocked by this issue and we hope this can be fixed for the next dune release. Thanks!

**Comment by josch** [comment]
Another update: we decided to work around this issue in dose3 by renaming `dose3.opam` to `dose3.opam2`. Probably nobody is using it right now anyways? According to https://opam.ocaml.org/packages/dose3/dose3.5.0.1/ only opam depends on dose3 and as far as I read the opam source, they do not depend on the `dose3.opam` module? A quick heads-up to @dra27 about the rename in case I assessed the situation wrongly.

**Comment by rgrinberg** [comment]
That's because we copy the opam file to `_build/install/default/lib/test/opam`.

I'm not sure how we can work around that.

**Comment by kit-ty-kate** [comment]
I'm curious, what is the reason behind installing the opam file?

**Comment by rgrinberg** [comment]
I know of no reason off the top of my head. Possibly @jeremiedimino knows

**Comment by dra27** [comment]
It could also be installed as `pkg.opam` which would eliminate the problem.

The opam file is installed because jbuilder adopted [odig’s metadata conventions](https://erratique.ch/software/odig/doc/packaging.html#meta), so changing the name would require a small change to odig (the change _might_ be good anyway, as odig could use the same conventions as opam for locating the opam file for a given package name)

**Comment by dra27** [comment]
Another possibility, not requiring changes to odig: I don't think that the directory a sub-package gets installed in is significant to anyone (the directory is mentioned both in `META` and `dune-package`, I think?). Dune could pick a different directory name in this case, given that although the _precise_ list of installed artefacts won't be known, the _total_ list of installed artefacts is. It's not the simplest change, but it should be possible to change the installation directory of a sub-package at the point at which the first `public_name` with the package name is encountered?

It's either not happened so far (or because it's in user control you'd just change the filename), but it there's no particular reason not to install any file `foo` as part of a package and also have a sub-package also called `foo`.

**Comment by jeremiedimino** [comment]
> The opam file is installed because jbuilder adopted odig’s metadata conventions, so changing the name would require a small change to odig (the change might be good anyway, as odig could use the same conventions as opam for locating the opam file for a given package name)

I confirm. 

I don't know if odig is widely used though; maybe it's not used very much and following odig's conventions is no longer relevant at all. Not sure.

**Comment by josch** [comment]
> Another possibility, not requiring changes to odig: I don't think that the directory a sub-package gets installed in is significant to anyone (the directory is mentioned both in `META` and `dune-package`, I think?). Dune could pick a different directory name in this case, given that although the _precise_ list of installed artefacts won't be known, the _total_ list of installed artefacts is. It's not the simplest change, but it should be possible to change the installation directory of a sub-package at the point at which the first `public_name` with the package name is encountered?
> 
> It's either not happened so far (or because it's in user control you'd just change the filename), but it there's no particular reason not to install any file `foo` as part of a package and also have a sub-package also called `foo`.

In dose3, which is affected by this issue, we already renamed our submodule "opam" to "opam2" to work around this issue. But we didn't make a new release yet. Could you elaborate a bit more how we should instruct dune such that we can still have a subpackage called "opam" and avoid the file conflict? Thanks!

**Comment by jeremiedimino** [comment]
One simple thing we can do right now to workaround this issue is not install the `opam` file when there is an `xxx.opam` library.

**Comment by dra27** [comment]
The workaround is nice. It's a shame if a library has to rename a package which pre-existed and we do expect to produce opam 3 within the next few years!

**Comment by josch** [comment]
Hi, I wanted to ask whether there is either a workaround for this issue that we can use with dune today or whether there is a rough ETA for when you think you can have a fix for this issue? We are currently planning the next release of dose3 and are wondering whether or not we should rename our "opam" module to "opam2" as a workaround. Even if dune introduces a fix, I don't think it would make sense for dose3 to then rename the module *again* back to "opam".

**Comment by rgrinberg** [comment]
I don’t think the team has any bandwidth to work on this now, but your own contribution would be welcome.
